proposals/accepted/047-standard-library-reform.rst

```diff
@@ -0,0 +1,559 @@
+
+Rust's ``cap-std``
+~~~~~~~~~~~~~~~~~~
+
+`cap-std <https://github.com/bytecodealliance/cap-std>`_ is a Rust library exploring what ergonomic IO interfaces for WASI system calls in a high level language should look like.
+On one hand, it is great, and we should borrow from it heavily.
+On the other hand, we should surpass it in not needing to be something on top of the "regular" standard library which ordinarily exposes more Unixy things than is appropriate.
+
+Rust's ``std`` and WASI
+~~~~~~~~~~~~~~~~~~~~~~~
+
+While the best experience comes from using ``cap-std`` as described above, Rust's ``std`` still makes sure to avoid indirecting through ``wasi-libc`` wherever possible.
+`This PR <https://github.com/rust-lang/rust/pull/63676>`_ made that change, using the ``wasi`` library (Rust bindings to WASI system calls) directly.
+This is what we should emulate in order to provide a top-tier programming environment for greenfield WebAssembly applications in Haskell.
+
+Prior attempts at splitting ``base``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There have been attempts to split ``base`` before, but they attempted to get everything done at once, setting a dangerously high bar for success.
+This approach here, by contrast, first and foremost seeks to avoid those difficulties and find a sustainable, suitably low-risk approach.
+It is much more concerned with how we safely approach these issues than what the exact outcome looks like.
+
+Technical Content
+-----------------
+
+Here is a plan to solve these issues.
+
+**Step 1A**: Task the CLC with defining new standard libraries
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Based on the conversation in `CLC Issue #015`_, ``base`` is exposing too much stuff, yet trying to limit what is exposed would be a big breaking change.
+
+The solution is to reach for another layer of indirection.
+The CLC should be tasked with devising new standard library interfaces, which would initially be implemented by reexporting modules from ``base``.
+
+The new library interfaces should be carefully designed in and of themselves to tackle many, but not all, of the issues above:
+
+- They should be designed *not* to break every release.
+  Even though the underlying ``base`` from which modules are reexported would continue to have its regular problematic major version bumps, the portion reexported should have very infrequent breaking changes.
+
+  This fixes **Problem 1**.
+
+- These libraries should be emphasized in all documentation, and users should be encouraged to use them and not ``base`` in new end-application code.
+  ``base``, in contrast, would be kept exposed as a mere legacy interface.
+  As code migrates over to use the new standard libraries, ``base`` should become less important.
+  GHC devs can therefore feel increasingly confident modifying parts of ``base`` which are *not* reexported in these new libraries.
```

Comment:
This doesn't solve the CLC 105 issue, because `base` is still considered part of their remit. It won't stop being that until it's not widely depended on. I think there's a chicken-egg issue to be addressed here in the proposal.